### PR TITLE
Impore the source sending feature

### PR DESCRIPTION
**Pull request by knightburton**
- Move the file handling into the extension file
- Create custom event and custom request
- Fix the after state of source sending

IoT.js-Debug-DCO-1.0-Signed-off-by: Imre Kiss kissi.szeged@partner.samsung.com

### CODE REVIEW

File: src/IotjsDebugger.ts [hunk 1]
```diff
@@ -349,6 +356,30 @@ class IotjsDebugSession extends LoggingDebugSession {
...
 
+  protected customRequest(command: string, response: DebugProtocol.Response, args: any): void {
+    this.log('customRequest');
+
+    switch (command) {
+      case 'sendSource': {
+        this._sourceSendingOptions.state = SOURCE_SENDING_STATES.IN_PROGRESS;
+        this._protocolhandler.sendClientSource(args.program.name, args.program.source)
+          .then(() => {
+            this.log('Source has been sended to the engine.');
```
Comment (tdusnoki) [inline src/IotjsDebugger.ts]: 'sent' instead of 'sended'

File: src/IotjsDebugger.ts [hunk 2]
```diff
@@ -17,21 +17,28 @@
...
 } from './JerryProtocolHandler';
-import { EVAL_RESULT_SUBTYPE } from './JerryProtocolConstants';
+import { EVAL_RESULT_SUBTYPE, CLIENT as CLIENT_PACKAGE } from './JerryProtocolConstants';
 
-class IotjsDebugSession extends LoggingDebugSession {
+enum SOURCE_SENDING_STATES {
+  NOP = 0,
+  WAITING = 1,
+  IN_PROGRESS = 2,
+  LAST_SENDED = 3
```
Comment (tdusnoki) [inline src/IotjsDebugger.ts]: 'SENT' instead of 'SENDED'

File: src/IotjsDebugger.ts [hunk 3]
```diff
@@ -140,23 +151,14 @@ class IotjsDebugSession extends LoggingDebugSession {
...
-            .then(() => this.log('Source has been sended to the engine.'))
-            .catch(error => {
-              this.sendErrorResponse(response, 0, error);
-            });
-        } else {
-          this.sendErrorResponse(response, 0, 'You must provide a valid path to source');
+      if (this._sourceSendingOptions.state === SOURCE_SENDING_STATES.NOP) {
+        this._sourceSendingOptions.state = SOURCE_SENDING_STATES.WAITING;
+        this.sendEvent(new Event('waitForSource'));
+      } else if (this._sourceSendingOptions.state === SOURCE_SENDING_STATES.LAST_SENDED) {
```
Comment (tdusnoki) [inline src/IotjsDebugger.ts]: 'SENT' instead of 'SENDED'

File: src/IotjsDebugger.ts [hunk 4]
```diff
@@ -349,6 +356,30 @@ class IotjsDebugSession extends LoggingDebugSession {
...
+  protected customRequest(command: string, response: DebugProtocol.Response, args: any): void {
+    this.log('customRequest');
+
+    switch (command) {
+      case 'sendSource': {
+        this._sourceSendingOptions.state = SOURCE_SENDING_STATES.IN_PROGRESS;
+        this._protocolhandler.sendClientSource(args.program.name, args.program.source)
+          .then(() => {
+            this.log('Source has been sended to the engine.');
+            this._sourceSendingOptions.state = SOURCE_SENDING_STATES.LAST_SENDED;
```
Comment (tdusnoki) [inline src/IotjsDebugger.ts]: 'SENT' instead of 'SENDED'

File: src/IotjsDebugger.ts (hunk 2 above)
Comment (knightburton) [reply src/IotjsDebugger.ts]: every, single, time.... :( sorry

**Pull request merged**